File: Freestylecoding.Math.FSharp.Tests/Natural/IncrementOperators.fs
namespace Natural

open Xunit
open Freestylecoding.Math

// NOTE: F# doesn't really have increment/decrement operators
//  It would mean the value is mutable
//  I'll get these tested in the C# tests
//	Maybe, after i confirm everything is working, I'll loop back and play with mutable to try it

//type public IncrementOperators() =
//    static member Increment( n:'T when 'T :> System.Numerics.IIncrementOperators<'T> ) : 'T =
//        'T.op_Increment( n )
//
//    [<Theory>]
//    [<InlineData( 0u,  1u )>]   // Sanity
//    [<InlineData( 1u,  2u )>]   // Sanity
//    [<InlineData( 9u, 10u )>]   // Sanity
//    member public this.Sanity value expected =
//        Assert.Equal(
//            Natural( [expected] ),
//            IncrementOperators.Increment( Natural( [value] ) )
//        )
//    
//    [<Fact>]
//    member public this.Overflow () =
//        Assert.Equal(
//            Natural( [1u; 0u] ),
//            IncrementOperators.Increment(
//                Natural( [0xFFFF_FFFFu] )
//            )
//        )
//        
//    [<Fact>]
//    member public this.CascadingOverflow () =
//        Assert.Equal(
//            Natural( [1u; 0u; 0u] ),
//            IncrementOperators.Increment(
//                Natural( [0xFFFF_FFFFu; 0xFFFF_FFFFu] )
//            )
//        )

// NOTE: The "Checked" tests are commentted out because they are not calling the correct operator
//  I cannot figure out who to force F# to call the correct one
//  We need to really make sure the C# side of the tests do call the right one

//type public CheckedIncrementOperators() =
//    static member CheckedIncrement( this:'T when 'T :> System.Numerics.IIncrementOperators<'T> ) : 'T =
//        'T.op_CheckedIncrement( this )
//
//    [<Theory>]
//    [<InlineData( 0u,  1u )>]   // Sanity
//    [<InlineData( 1u,  2u )>]   // Sanity
//    [<InlineData( 9u, 10u )>]   // Sanity
//    member public this.Sanity value expected =
//        Assert.Equal(
//            Natural( [expected] ),
//            CheckedIncrementOperators.CheckedIncrement( Natural( [value] ) )
//        )
//    
//    [<Fact>]
//    member public this.Overflow () =
//        Assert.Equal(
//            Natural( [1u; 0u] ),
//            CheckedIncrementOperators.CheckedIncrement(
//                Natural( [0xFFFF_FFFFu] )
//            )
//        )
//        
//    [<Fact>]
//    member public this.CascadingOverflow () =
//        Assert.Equal(
//            Natural( [1u; 0u; 0u] ),
//            CheckedIncrementOperators.CheckedIncrement(
//                Natural( [0xFFFF_FFFFu; 0xFFFF_FFFFu] )
//            )
//        )

File: Freestylecoding.Math.CSharp.Tests/Natural/IncrementOperators.cs
using System.Numerics;
using Xunit;

namespace Natural.Tests;

using static Freestylecoding.Math.CSharp.Tests.Helpers;
using Natural = Freestylecoding.Math.Natural;

public class IncrementOperators {
	private static void AssertIncrement<T>( T expected, T actual ) where T : IIncrementOperators<T> {
		// This one checks the return of the Increment expression
		Assert.Equal<T>( expected, ++actual );
		// This one ensures the original was mutated
		Assert.Equal<T>( expected, actual );
	}

	[Theory]
	[InlineData( 0u, 1u )]   // Sanity
	[InlineData( 1u, 2u )]   // Sanity
	[InlineData( 9u, 10u )]   // Sanity
	public void Sanity( uint value, uint expected ) =>
		AssertIncrement(
			new Natural( expected ),
			value
		);

	[Fact]
	public void Overflow() =>
		AssertIncrement(
			FListNatural( 1u, 0u ),
			new Natural( 0xFFFF_FFFFu )
		);

	[Fact]
	public void CascadingOverflow() =>
		AssertIncrement(
			FListNatural( 1u, 0u, 0u ),
			FListNatural( 0xFFFF_FFFFu, 0xFFFF_FFFFu )
		);

	[Fact( Skip = "Postfix won't work until Natural is a ValueType" )]
	public void Postfix() {
		Natural actual = Natural.Zero;
		Assert.Equal(
			Natural.Zero,
			actual++
		);
		Assert.Equal(
			Natural.One,
			actual
		);
	}
}

public class CheckedIncrementOperators {
	private static void AssertCheckedIncrement<T>( T expected, T actual ) where T : IIncrementOperators<T> {
		checked {
			// This one checks the return of the Increment expression
			Assert.Equal<T>( expected, ++actual );
			// This one ensures the original was mutated
			Assert.Equal<T>( expected, actual );
		}
	}

	[Theory]
	[InlineData( 0u, 1u )]   // Sanity
	[InlineData( 1u, 2u )]   // Sanity
	[InlineData( 9u, 10u )]   // Sanity
	public void Sanity( uint value, uint expected ) =>
		AssertCheckedIncrement(
			new Natural( expected ),
			value
		);

	[Fact]
	public void Overflow() =>
		AssertCheckedIncrement(
			FListNatural( 1u, 0u ),
			new Natural( 0xFFFF_FFFFu )
		);

	[Fact]
	public void CascadingOverflow() =>
		AssertCheckedIncrement(
			FListNatural( 1u, 0u, 0u ),
			FListNatural( 0xFFFF_FFFFu, 0xFFFF_FFFFu )
		);

	[Fact( Skip = "Postfix won't work until Natural is a ValueType" )]
	public void Postfix() {
		Natural actual = Natural.Zero;
		Assert.Equal(
			Natural.Zero,
			checked( actual++ )
		);
		Assert.Equal(
			Natural.One,
			actual
		);
	}
}